File: services/iceberg-catalog-service/src/metrics.rs
//! Prometheus metrics for `iceberg-catalog-service`.
//!
//! The names below match the closing-task spec § 9 and are stable
//! across deployments. Each is a process-global `Lazy` so handlers can
//! `.with_label_values(...).inc()` without threading the registry.

use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use once_cell::sync::Lazy;
use prometheus::{
    Encoder, IntCounterVec, IntGaugeVec, Opts, TextEncoder, register_int_counter_vec,
    register_int_gauge_vec,
};

pub static REST_REQUESTS_TOTAL: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        Opts::new(
            "iceberg_rest_catalog_requests_total",
            "REST Catalog requests by method, endpoint and HTTP status"
        ),
        &["method", "endpoint", "status"]
    )
    .expect("register iceberg_rest_catalog_requests_total")
});

pub static OAUTH_TOKENS_ISSUED: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        Opts::new(
            "iceberg_oauth_token_issued_total",
            "OAuth2 tokens issued by grant_type"
        ),
        &["grant_type"]
    )
    .expect("register iceberg_oauth_token_issued_total")
});

pub static TABLES_TOTAL: Lazy<IntGaugeVec> = Lazy::new(|| {
    register_int_gauge_vec!(
        Opts::new(
            "iceberg_tables_total",
            "Number of Iceberg tables tracked by the catalog by format_version"
        ),
        &["format_version"]
    )
    .expect("register iceberg_tables_total")
});

pub static METADATA_FILES_TOTAL: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        Opts::new(
            "iceberg_metadata_files_total",
            "Cumulative count of v{N}.metadata.json files written"
        ),
        &["table_uuid"]
    )
    .expect("register iceberg_metadata_files_total")
});

// ─── P2 metrics ───────────────────────────────────────────────────────

pub static COMMIT_CONFLICTS_TOTAL: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        Opts::new(
            "iceberg_commit_conflicts_total",
            "Multi-table commit conflicts surfaced as 409 Retryable to the build executor"
        ),
        &["conflicting_with"]
    )
    .expect("register iceberg_commit_conflicts_total")
});

pub static SCHEMA_STRICT_REJECTIONS_TOTAL: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        Opts::new(
            "iceberg_schema_strict_rejections_total",
            "Commits rejected by strict-mode schema enforcement (per delta kind)"
        ),
        &["delta_kind"]
    )
    .expect("register iceberg_schema_strict_rejections_total")
});

pub static BRANCH_ALIAS_APPLIED_TOTAL: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        Opts::new(
            "iceberg_branch_alias_applied_total",
            "Master/main alias rewrites applied per Foundry doc § Default branches"
        ),
        &["from", "to"]
    )
    .expect("register iceberg_branch_alias_applied_total")
});

pub static FOUNDRY_TRANSACTIONS_TOTAL: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        Opts::new(
            "iceberg_foundry_transactions_total",
            "FoundryIcebergTxn lifecycle counter (begin/commit/abort)"
        ),
        &["lifecycle"]
    )
    .expect("register iceberg_foundry_transactions_total")
});

pub fn record_rest_request(method: &str, endpoint: &str, status: u16) {
    REST_REQUESTS_TOTAL
        .with_label_values(&[method, endpoint, &status.to_string()])
        .inc();
}

/// Prometheus exposition endpoint (`GET /metrics`). Renders the global
/// registry in the standard text format.
pub async fn render_metrics() -> Response {
    let encoder = TextEncoder::new();
    let metric_families = prometheus::gather();
    let mut buffer = Vec::new();
    if let Err(error) = encoder.encode(&metric_families, &mut buffer) {
        tracing::error!(?error, "failed to encode prometheus metrics");
        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
    }

    (
        StatusCode::OK,
        [(
            axum::http::header::CONTENT_TYPE,
            encoder.format_type().to_string(),
        )],
        buffer,
    )
        .into_response()
}

File: services/iceberg-catalog-service/internal/metrics/metrics.go
// Package metrics declares the Prometheus families specific to
// iceberg-catalog-service. Mirrors services/iceberg-catalog-service/src/metrics.rs
// — names, labels and help text are byte-for-byte identical so the
// dashboards and alerts wired against the Rust crate keep firing
// against the Go service unchanged.
//
// The HTTP instrumentation families (RESTRequestsTotal,
// RESTRequestLatencySeconds, RESTRequestsInFlight) are populated by the
// chi middleware in package server. The remaining families
// (OAuth tokens, tables-by-format, commit conflicts, etc.) are surfaced
// directly from the relevant handlers.
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"

	"github.com/openfoundry/openfoundry-go/libs/observability"
)

// Metrics owns the typed Prometheus handles. Instantiated once in
// cmd/iceberg-catalog-service/main.go and threaded through the server
// Deps so handlers and middleware share the same registered families.
type Metrics struct {
	// RESTRequestsTotal counts every request handled by the chi router,
	// labelled by method, route pattern and HTTP status. Mirrors
	// `iceberg_rest_catalog_requests_total` from the Rust crate.
	RESTRequestsTotal *prometheus.CounterVec
	// RESTRequestLatencySeconds is the per-route latency histogram.
	// Buckets follow the platform default (5ms…10s, geometric).
	RESTRequestLatencySeconds *prometheus.HistogramVec
	// RESTRequestsInFlight is a gauge of currently-handled requests
	// per route — useful to spot saturation independent of latency.
	RESTRequestsInFlight *prometheus.GaugeVec

	// OAuthTokensIssued counts OAuth2 tokens minted, by grant_type.
	OAuthTokensIssued *prometheus.CounterVec
	// TablesTotal tracks the live count of catalog tables, by
	// format-version. Set by the table CRUD handlers.
	TablesTotal *prometheus.GaugeVec
	// MetadataFilesTotal counts metadata.json files written, per table.
	MetadataFilesTotal *prometheus.CounterVec

	// CommitConflictsTotal counts multi-table commit conflicts surfaced
	// as HTTP 409 to the build executor.
	CommitConflictsTotal *prometheus.CounterVec
	// SchemaStrictRejectionsTotal counts commits rejected by the
	// strict-mode schema enforcer, per delta kind.
	SchemaStrictRejectionsTotal *prometheus.CounterVec
	// BranchAliasAppliedTotal counts master/main alias rewrites that
	// the branch resolver applies on read or write.
	BranchAliasAppliedTotal *prometheus.CounterVec
	// FoundryTransactionsTotal is the FoundryIcebergTxn lifecycle
	// counter (begin/commit/abort).
	FoundryTransactionsTotal *prometheus.CounterVec
}

// New registers the families on the supplied observability registry
// and returns the typed handle.
func New(o *observability.Metrics) *Metrics {
	m := &Metrics{
		RESTRequestsTotal: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "iceberg_rest_catalog_requests_total",
				Help: "REST Catalog requests by method, endpoint and HTTP status",
			},
			[]string{"method", "endpoint", "status"},
		),
		RESTRequestLatencySeconds: prometheus.NewHistogramVec(
			prometheus.HistogramOpts{
				Name:    "iceberg_rest_catalog_request_latency_seconds",
				Help:    "REST Catalog request latency in seconds, by method and endpoint",
				Buckets: prometheus.ExponentialBuckets(0.005, 2, 12),
			},
			[]string{"method", "endpoint"},
		),
		RESTRequestsInFlight: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "iceberg_rest_catalog_requests_in_flight",
				Help: "REST Catalog requests currently being handled, by method and endpoint",
			},
			[]string{"method", "endpoint"},
		),
		OAuthTokensIssued: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "iceberg_oauth_token_issued_total",
				Help: "OAuth2 tokens issued by grant_type",
			},
			[]string{"grant_type"},
		),
		TablesTotal: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "iceberg_tables_total",
				Help: "Number of Iceberg tables tracked by the catalog by format_version",
			},
			[]string{"format_version"},
		),
		MetadataFilesTotal: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "iceberg_metadata_files_total",
				Help: "Cumulative count of v{N}.metadata.json files written",
			},
			[]string{"table_uuid"},
		),
		CommitConflictsTotal: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "iceberg_commit_conflicts_total",
				Help: "Multi-table commit conflicts surfaced as 409 Retryable to the build executor",
			},
			[]string{"conflicting_with"},
		),
		SchemaStrictRejectionsTotal: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "iceberg_schema_strict_rejections_total",
				Help: "Commits rejected by strict-mode schema enforcement (per delta kind)",
			},
			[]string{"delta_kind"},
		),
		BranchAliasAppliedTotal: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "iceberg_branch_alias_applied_total",
				Help: "Master/main alias rewrites applied per Foundry doc § Default branches",
			},
			[]string{"from", "to"},
		),
		FoundryTransactionsTotal: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "iceberg_foundry_transactions_total",
				Help: "FoundryIcebergTxn lifecycle counter (begin/commit/abort)",
			},
			[]string{"lifecycle"},
		),
	}
	o.Register(m.RESTRequestsTotal)
	o.Register(m.RESTRequestLatencySeconds)
	o.Register(m.RESTRequestsInFlight)
	o.Register(m.OAuthTokensIssued)
	o.Register(m.TablesTotal)
	o.Register(m.MetadataFilesTotal)
	o.Register(m.CommitConflictsTotal)
	o.Register(m.SchemaStrictRejectionsTotal)
	o.Register(m.BranchAliasAppliedTotal)
	o.Register(m.FoundryTransactionsTotal)
	return m
}
